File: clients/python/tests/test_consumer.py
# Copyright 2026 Nikolay Samokhvalov. Apache-2.0 license.

"""``Consumer`` end-to-end: dispatch, unknown-type handling, error -> nack."""

import logging
import threading
import time
from unittest import mock
from unittest.mock import MagicMock

import pgque


def _run_consumer_for(consumer: pgque.Consumer, seconds: float) -> threading.Thread:
    """Start a consumer in a background thread, stop it after `seconds`."""
    t = threading.Thread(target=consumer.start, daemon=True)
    t.start()

    def _stopper():
        time.sleep(seconds)
        consumer.stop()

    threading.Thread(target=_stopper, daemon=True).start()
    return t


def test_consumer_default_max_messages_requests_whole_batch(dsn):
    cons = pgque.Consumer(dsn=dsn, queue="q", name="c")
    assert cons.max_messages == 2_147_483_647


def test_consumer_configured_max_messages_is_preserved(dsn):
    cons = pgque.Consumer(dsn=dsn, queue="q", name="c", max_messages=123)
    assert cons.max_messages == 123


def test_consumer_poll_once_passes_default_max_messages(dsn):
    cons = pgque.Consumer(dsn=dsn, queue="q", name="c")
    fake_conn = MagicMock()

    with mock.patch.object(pgque.PgqueClient, "receive", return_value=[]) as receive:
        cons._poll_once(fake_conn)

    receive.assert_called_once_with("q", "c", 2_147_483_647)


def test_consumer_poll_once_passes_configured_max_messages(dsn):
    cons = pgque.Consumer(dsn=dsn, queue="q", name="c", max_messages=123)
    fake_conn = MagicMock()

    with mock.patch.object(pgque.PgqueClient, "receive", return_value=[]) as receive:
        cons._poll_once(fake_conn)

    receive.assert_called_once_with("q", "c", 123)


def test_consumer_dispatches_by_event_type(dsn, conn, setup_queue):
    queue, consumer_name = setup_queue
    client = pgque.PgqueClient(conn)
    client.send(queue, {"i": 1}, type="evt.a")
    client.send(queue, {"i": 2}, type="evt.b")
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()

    seen_a: list = []
    seen_b: list = []
    cons = pgque.Consumer(
        dsn=dsn, queue=queue, name=consumer_name, poll_interval=1
    )

    @cons.on("evt.a")
    def _a(m: pgque.Message):
        seen_a.append(m.payload)

    @cons.on("evt.b")
    def _b(m: pgque.Message):
        seen_b.append(m.payload)

    t = _run_consumer_for(cons, 3.0)
    t.join(timeout=5.0)

    assert len(seen_a) == 1
    assert len(seen_b) == 1


def test_consumer_default_handler_catches_unknown(dsn, conn, setup_queue):
    queue, consumer_name = setup_queue
    client = pgque.PgqueClient(conn)
    client.send(queue, {"x": 99}, type="never.registered.type")
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()

    fallback: list = []
    cons = pgque.Consumer(
        dsn=dsn, queue=queue, name=consumer_name, poll_interval=1
    )

    @cons.on("*")
    def _default(m: pgque.Message):
        fallback.append(m)

    t = _run_consumer_for(cons, 3.0)
    t.join(timeout=5.0)

    assert len(fallback) == 1
    assert fallback[0].type == "never.registered.type"


def test_consumer_nacks_on_handler_error(dsn, conn, setup_queue):
    queue, consumer_name = setup_queue
    client = pgque.PgqueClient(conn)
    client.send(queue, {"i": 1}, type="evt.fail")
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()

    calls = {"n": 0}
    cons = pgque.Consumer(
        dsn=dsn, queue=queue, name=consumer_name,
        poll_interval=1, retry_after=0,
    )

    @cons.on("evt.fail")
    def _boom(m: pgque.Message):
        calls["n"] += 1
        raise RuntimeError("simulated failure")

    t = _run_consumer_for(cons, 3.0)
    t.join(timeout=5.0)

    # The handler ran at least once, and the failing message landed in
    # the retry queue (not silently dropped).
    assert calls["n"] >= 1
    cnt = conn.execute(
        "select count(*) from pgque.retry_queue rq "
        "join pgque.queue q on q.queue_id = rq.ev_queue "
        "where q.queue_name = %s",
        (queue,),
    ).fetchone()[0]
    assert cnt >= 1


def _retry_count_for_msg(conn, queue: str, msg_id: int) -> int:
    return conn.execute(
        "select count(*) from pgque.retry_queue rq "
        "join pgque.queue q on q.queue_id = rq.ev_queue "
        "where q.queue_name = %s and rq.ev_id = %s",
        (queue, msg_id),
    ).fetchone()[0]


def _dead_letter_count_for_msg(conn, queue: str, msg_id: int) -> int:
    return conn.execute(
        "select count(*) from pgque.dead_letter dl "
        "join pgque.queue q on q.queue_id = dl.dl_queue_id "
        "where q.queue_name = %s and dl.ev_id = %s",
        (queue, msg_id),
    ).fetchone()[0]


def test_consumer_nacks_unhandled_event_type(dsn, conn, setup_queue):
    """Default behavior: unhandled type is nacked.

    The message must land in retry_queue (or dead_letter if max_retries=0),
    and a follow-up receive must not return the same msg_id, proving the
    batch advanced.
    """
    queue, consumer_name = setup_queue
    client = pgque.PgqueClient(conn)
    msg_id = client.send(queue, {"x": 1}, type="totally.unregistered.type")
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()

    # Consumer with NO handler for "totally.unregistered.type"
    # and NO default handler either.
    cons = pgque.Consumer(
        dsn=dsn, queue=queue, name=consumer_name, poll_interval=1
    )

    t = _run_consumer_for(cons, 3.0)
    t.join(timeout=5.0)

    # Must be routed to retry_queue (queue_max_retries is unset by default,
    # so the message is retried rather than dead-lettered).
    rq = _retry_count_for_msg(conn, queue, msg_id)
    dlq = _dead_letter_count_for_msg(conn, queue, msg_id)
    assert rq + dlq >= 1, (
        f"unhandled event was not nacked: retry_queue={rq} dead_letter={dlq}"
    )

    # The batch advanced: a fresh receive must not return the same msg_id.
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()
    follow_up = client.receive(queue, consumer_name, max_messages=10)
    assert all(m.msg_id != msg_id for m in follow_up), (
        "batch did not advance past unhandled msg_id; got it again on receive"
    )


def test_consumer_acks_unhandled_event_type_when_opt_in(
    dsn, conn, setup_queue, caplog
):
    """Opt-in: unknown_handler='ack' restores warn+ack semantics.

    Must NOT appear in retry_queue, must log a WARNING, and a follow-up
    receive must not return the same msg_id.
    """
    queue, consumer_name = setup_queue
    client = pgque.PgqueClient(conn)
    msg_id = client.send(queue, {"x": 1}, type="totally.unregistered.type")
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()

    cons = pgque.Consumer(
        dsn=dsn,
        queue=queue,
        name=consumer_name,
        poll_interval=1,
        unknown_handler="ack",
    )

    with caplog.at_level(logging.WARNING, logger="pgque"):
        t = _run_consumer_for(cons, 3.0)
        t.join(timeout=5.0)

    # Must NOT be in retry_queue or dead_letter -- it was acked, not nacked.
    rq = _retry_count_for_msg(conn, queue, msg_id)
    dlq = _dead_letter_count_for_msg(conn, queue, msg_id)
    assert rq == 0, (
        "unhandled event type was nacked (found in retry_queue); expected ack"
    )
    assert dlq == 0, (
        "unhandled event type was dead-lettered; expected ack"
    )

    # A WARNING must have been logged containing the type.
    warning_lines = [
        r.getMessage() for r in caplog.records if r.levelno == logging.WARNING
    ]
    assert any("totally.unregistered.type" in m for m in warning_lines), (
        "expected a WARNING mentioning the unhandled event type"
    )

    # The batch advanced: a fresh receive must not return the same msg_id.
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()
    follow_up = client.receive(queue, consumer_name, max_messages=10)
    assert all(m.msg_id != msg_id for m in follow_up), (
        "batch did not advance past unhandled msg_id; got it again on receive"
    )


def test_consumer_rejects_invalid_unknown_handler(dsn, setup_queue):
    """Constructor must reject values other than 'nack' / 'ack'."""
    queue, consumer_name = setup_queue
    try:
        pgque.Consumer(
            dsn=dsn,
            queue=queue,
            name=consumer_name,
            unknown_handler="bogus",  # type: ignore[arg-type]
        )
    except ValueError:
        return
    raise AssertionError("expected ValueError for invalid unknown_handler")


def test_consumer_stop_returns_promptly(dsn, setup_queue):
    queue, consumer_name = setup_queue
    cons = pgque.Consumer(
        dsn=dsn, queue=queue, name=consumer_name, poll_interval=10
    )
    t = threading.Thread(target=cons.start, daemon=True)
    t.start()
    time.sleep(0.5)  # let it enter the loop
    cons.stop()
    t.join(timeout=15)
    assert not t.is_alive(), "consumer did not stop after stop()"


def test_consumer_does_not_ack_when_unknown_type_nack_fails(
    dsn, conn, setup_queue
):
    """If ``nack()`` raises in the unknown-handler path, the batch must
    NOT be acked. PgQ must redeliver the whole batch on the next poll.
    """
    queue, consumer_name = setup_queue
    client = pgque.PgqueClient(conn)
    msg_id = client.send(queue, {"x": 1}, type="totally.unregistered.type")
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()

    cons = pgque.Consumer(
        dsn=dsn, queue=queue, name=consumer_name, poll_interval=1
    )

    real_client_init = pgque.PgqueClient.__init__
    ack_calls: list[int] = []
    nack_calls: list[int] = []

    def fake_init(self, c):
        real_client_init(self, c)
        original_ack = self.ack
        original_nack = self.nack

        def spy_ack(batch_id):
            ack_calls.append(batch_id)
            return original_ack(batch_id)

        def explode_nack(batch_id, msg, retry_after=60, reason=None):
            nack_calls.append(msg.msg_id)
            raise RuntimeError("simulated nack failure")

        self.ack = spy_ack  # type: ignore[method-assign]
        self.nack = explode_nack  # type: ignore[method-assign]

    with mock.patch.object(pgque.PgqueClient, "__init__", fake_init):
        t = _run_consumer_for(cons, 3.0)
        t.join(timeout=5.0)

    assert nack_calls, "nack was never called for the unhandled message"
    assert ack_calls == [], (
        f"ack must not be called when nack raised; got ack_calls={ack_calls}"
    )

    # Message must still be visible: re-receive returns the same msg_id.
    follow_up = client.receive(queue, consumer_name, max_messages=10)
    assert any(m.msg_id == msg_id for m in follow_up), (
        "batch advanced even though nack failed; data was lost"
    )
    # Cleanup: ack the redelivered batch so the queue tear-down is clean.
    if follow_up:
        client.ack(follow_up[0].batch_id)
        conn.commit()


def test_consumer_does_not_ack_when_handler_error_nack_fails(
    dsn, conn, setup_queue
):
    """If ``nack()`` raises in the handler-error path, the batch must
    NOT be acked.
    """
    queue, consumer_name = setup_queue
    client = pgque.PgqueClient(conn)
    msg_id = client.send(queue, {"i": 1}, type="evt.fail")
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()

    cons = pgque.Consumer(
        dsn=dsn, queue=queue, name=consumer_name,
        poll_interval=1, retry_after=0,
    )

    @cons.on("evt.fail")
    def _boom(m: pgque.Message):
        raise RuntimeError("handler boom")

    real_client_init = pgque.PgqueClient.__init__
    ack_calls: list[int] = []
    nack_calls: list[int] = []

    def fake_init(self, c):
        real_client_init(self, c)
        original_ack = self.ack

        def spy_ack(batch_id):
            ack_calls.append(batch_id)
            return original_ack(batch_id)

        def explode_nack(batch_id, msg, retry_after=60, reason=None):
            nack_calls.append(msg.msg_id)
            raise RuntimeError("simulated nack failure")

        self.ack = spy_ack  # type: ignore[method-assign]
        self.nack = explode_nack  # type: ignore[method-assign]

    with mock.patch.object(pgque.PgqueClient, "__init__", fake_init):
        t = _run_consumer_for(cons, 3.0)
        t.join(timeout=5.0)

    assert nack_calls, "nack was never called after handler raised"
    assert ack_calls == [], (
        f"ack must not be called when nack raised; got ack_calls={ack_calls}"
    )

    follow_up = client.receive(queue, consumer_name, max_messages=10)
    assert any(m.msg_id == msg_id for m in follow_up), (
        "batch advanced even though nack failed; data was lost"
    )
    if follow_up:
        client.ack(follow_up[0].batch_id)
        conn.commit()


def test_consumer_stop_returns_within_2s_while_waiting(dsn, setup_queue):
    """``stop()`` must unblock the LISTEN/NOTIFY wait promptly, even when
    no notifications arrive and ``poll_interval`` is large.
    """
    queue, consumer_name = setup_queue
    cons = pgque.Consumer(
        dsn=dsn, queue=queue, name=consumer_name, poll_interval=30
    )
    t = threading.Thread(target=cons.start, daemon=True)
    t.start()
    # Let the consumer enter the LISTEN wait (after the first empty poll).
    time.sleep(1.0)

    t0 = time.monotonic()
    cons.stop()
    t.join(timeout=5.0)
    elapsed = time.monotonic() - t0

    assert not t.is_alive(), "consumer thread did not stop"
    assert elapsed < 2.0, (
        f"stop() took {elapsed:.2f}s; expected prompt (<2s) shutdown"
    )


def test_consumer_wakes_on_pg_notify_before_poll_interval(
    dsn, conn, setup_queue
):
    """A real ``pg_notify('pgque_<queue>', ...)`` from a producer thread
    must wake the consumer well before ``poll_interval`` elapses.
    """
    queue, consumer_name = setup_queue
    client = pgque.PgqueClient(conn)

    received: list[pgque.Message] = []
    received_evt = threading.Event()

    cons = pgque.Consumer(
        dsn=dsn, queue=queue, name=consumer_name, poll_interval=30
    )

    @cons.on("evt.wake")
    def _on(m: pgque.Message):
        received.append(m)
        received_evt.set()

    t = threading.Thread(target=cons.start, daemon=True)
    t.start()
    # Allow the consumer to drain its first (empty) poll and enter the wait.
    time.sleep(1.5)

    # Send + tick + NOTIFY from the test thread.
    t_send = time.monotonic()
    client.send(queue, {"i": 1}, type="evt.wake")
    conn.commit()
    conn.execute("select pgque.force_tick(%s)", (queue,))
    conn.execute("select pgque.ticker()")
    conn.commit()
    conn.execute(f"notify pgque_{queue}, 'go'")
    conn.commit()

    woke = received_evt.wait(timeout=5.0)
    elapsed = time.monotonic() - t_send

    cons.stop()
    t.join(timeout=5.0)

    assert woke, "consumer did not wake on pg_notify within 5s"
    assert len(received) == 1
    # poll_interval=30s, so anything well under that proves the wait woke
    # on the notification rather than the polling timer.
    assert elapsed < 5.0, (
        f"consumer woke too slowly ({elapsed:.2f}s); LISTEN/NOTIFY did not "
        f"unblock the wait"
    )

File: clients/ruby/test/test_consumer.rb
# Copyright 2026 Nikolay Samokhvalov. Apache-2.0 license.

require_relative "test_helper"
require "logger"
require "stringio"

class TestConsumerUnit < Minitest::Test
  include PgqueTest::Helpers

  class FakeTxConn
    def transaction
      yield self
    end
  end

  class SpyClient
    attr_reader :receive_calls

    def initialize(*)
      @receive_calls = []
    end

    def receive(queue, consumer, max_messages)
      @receive_calls << [queue, consumer, max_messages]
      []
    end

    def ack(_batch_id)
      1
    end

    def nack(*); end
  end

  def test_consumer_default_max_messages_requests_whole_batch
    cons = Pgque::Consumer.new(dsn, queue: "q", name: "c")
    assert_equal 2_147_483_647, cons.max_messages
  end

  def test_consumer_configured_max_messages_is_preserved
    cons = Pgque::Consumer.new(dsn, queue: "q", name: "c", max_messages: 123)
    assert_equal 123, cons.max_messages
  end

  def test_consumer_poll_once_passes_default_max_messages
    cons = Pgque::Consumer.new(dsn, queue: "q", name: "c")
    spy = SpyClient.new
    Pgque::Client.stub :new, ->(*) { spy } do
      cons.poll_once(FakeTxConn.new)
    end
    assert_equal [["q", "c", 2_147_483_647]], spy.receive_calls
  end

  def test_consumer_poll_once_passes_configured_max_messages
    cons = Pgque::Consumer.new(dsn, queue: "q", name: "c", max_messages: 123)
    spy = SpyClient.new
    Pgque::Client.stub :new, ->(*) { spy } do
      cons.poll_once(FakeTxConn.new)
    end
    assert_equal [["q", "c", 123]], spy.receive_calls
  end

  def test_consumer_rejects_invalid_unknown_handler_policy
    skip_dsn_for_this_class!
    assert_raises(ArgumentError) do
      Pgque::Consumer.new("dummy", queue: "q", name: "c",
                          unknown_handler_policy: "bogus")
    end
  end

  def test_consumer_dead_interval_without_subconsumer_raises
    skip_dsn_for_this_class!
    assert_raises(ArgumentError) do
      Pgque::Consumer.new("dummy", queue: "q", name: "c",
                          dead_interval: "5 minutes")
    end
  end

  def test_invalid_pgque_log_level_falls_back_to_fatal
    skip_dsn_for_this_class!
    old = ENV["PGQUE_LOG_LEVEL"]
    ENV["PGQUE_LOG_LEVEL"] = " warning "

    cons = Pgque::Consumer.new("dummy", queue: "q", name: "c")
    assert_equal Logger::FATAL, cons.logger.level
  ensure
    ENV["PGQUE_LOG_LEVEL"] = old
  end

  private

  # Some unit tests don't actually connect; allow them even without DSN.
  def skip_dsn_for_this_class!
    # The setup-level skip already passes when DSN is set; this method is
    # here so the structure is symmetric with the integration tests.
  end
end

class TestConsumerIntegration < Minitest::Test
  include PgqueTest::Helpers

  def run_consumer_for(consumer, seconds)
    t = Thread.new { consumer.start }
    Thread.new do
      sleep seconds
      consumer.stop
    end
    t
  end

  def force_tick(conn, queue)
    conn.exec_params("select pgque.force_next_tick($1)", [queue])
    conn.exec_params("select pgque.ticker($1)", [queue])
  end

  def silent_logger
    log = Logger.new(StringIO.new)
    log.level = Logger::FATAL
    log
  end

  def capturing_logger
    io = StringIO.new
    log = Logger.new(io)
    log.level = Logger::WARN
    [log, io]
  end

  def retry_count_for_msg(conn, queue, msg_id)
    conn.exec_params(
      "select count(*) from pgque.retry_queue rq " \
      "join pgque.queue q on q.queue_id = rq.ev_queue " \
      "where q.queue_name = $1 and rq.ev_id = $2",
      [queue, msg_id],
    ).values[0][0].to_i
  end

  def dlq_count_for_msg(conn, queue, msg_id)
    conn.exec_params(
      "select count(*) from pgque.dead_letter dl " \
      "join pgque.queue q on q.queue_id = dl.dl_queue_id " \
      "where q.queue_name = $1 and dl.ev_id = $2",
      [queue, msg_id],
    ).values[0][0].to_i
  end

  def test_consumer_dispatches_by_event_type
    with_queue do |queue, consumer_n, conn|
      client = Pgque::Client.new(conn)
      client.send(queue, { "i" => 1 }, type: "evt.a")
      client.send(queue, { "i" => 2 }, type: "evt.b")
      force_tick(conn, queue)

      seen_a = []
      seen_b = []
      cons = Pgque::Consumer.new(dsn, queue: queue, name: consumer_n,
                                 poll_interval: 1, logger: silent_logger)
      cons.on("evt.a") { |m| seen_a << m.payload }
      cons.on("evt.b") { |m| seen_b << m.payload }

      run_consumer_for(cons, 3.0).join(5.0)

      assert_equal 1, seen_a.size
      assert_equal 1, seen_b.size
    end
  end

  def test_consumer_default_handler_catches_unknown
    with_queue do |queue, consumer_n, conn|
      client = Pgque::Client.new(conn)
      client.send(queue, { "x" => 99 }, type: "never.registered.type")
      force_tick(conn, queue)

      fallback = []
      cons = Pgque::Consumer.new(dsn, queue: queue, name: consumer_n,
                                 poll_interval: 1, logger: silent_logger)
      cons.on("*") { |m| fallback << m }

      run_consumer_for(cons, 3.0).join(5.0)

      assert_equal 1, fallback.size
      assert_equal "never.registered.type", fallback[0].type
    end
  end

  def test_consumer_nacks_on_handler_error
    with_queue do |queue, consumer_n, conn|
      client = Pgque::Client.new(conn)
      client.send(queue, { "i" => 1 }, type: "evt.fail")
      force_tick(conn, queue)

      n_calls = 0
      cons = Pgque::Consumer.new(dsn, queue: queue, name: consumer_n,
                                 poll_interval: 1, retry_after: 0,
                                 logger: silent_logger)
      cons.on("evt.fail") { |_m| n_calls += 1; raise "simulated failure" }

      run_consumer_for(cons, 3.0).join(5.0)

      assert_operator n_calls, :>=, 1
      cnt = conn.exec_params(
        "select count(*) from pgque.retry_queue rq " \
        "join pgque.queue q on q.queue_id = rq.ev_queue " \
        "where q.queue_name = $1",
        [queue],
      ).values[0][0].to_i
      assert_operator cnt, :>=, 1
    end
  end

  def test_consumer_nacks_unhandled_event_type
    with_queue do |queue, consumer_n, conn|
      client = Pgque::Client.new(conn)
      msg_id = client.send(queue, { "x" => 1 }, type: "totally.unregistered.type")
      force_tick(conn, queue)

      cons = Pgque::Consumer.new(dsn, queue: queue, name: consumer_n,
                                 poll_interval: 1, logger: silent_logger)
      run_consumer_for(cons, 3.0).join(5.0)

      rq = retry_count_for_msg(conn, queue, msg_id)
      dlq = dlq_count_for_msg(conn, queue, msg_id)
      assert_operator rq + dlq, :>=, 1,
                      "unhandled event was not nacked: rq=#{rq} dlq=#{dlq}"

      force_tick(conn, queue)
      follow = client.receive(queue, consumer_n, 10)
      refute(follow.any? { |m| m.msg_id == msg_id },
             "batch did not advance past unhandled msg_id")
      client.ack(follow[0].batch_id) if follow.any?
    end
  end

  def test_consumer_acks_unhandled_event_type_when_opt_in
    with_queue do |queue, consumer_n, conn|
      client = Pgque::Client.new(conn)
      msg_id = client.send(queue, { "x" => 1 }, type: "totally.unregistered.type")
      force_tick(conn, queue)

      log, io = capturing_logger
      cons = Pgque::Consumer.new(dsn, queue: queue, name: consumer_n,
                                 poll_interval: 1,
                                 unknown_handler_policy: "ack",
                                 logger: log)
      run_consumer_for(cons, 3.0).join(5.0)

      assert_equal 0, retry_count_for_msg(conn, queue, msg_id)
      assert_equal 0, dlq_count_for_msg(conn, queue, msg_id)
      assert_includes io.string, "totally.unregistered.type"

      force_tick(conn, queue)
      follow = client.receive(queue, consumer_n, 10)
      refute(follow.any? { |m| m.msg_id == msg_id })
      client.ack(follow[0].batch_id) if follow.any?
    end
  end

  def test_consumer_stop_returns_promptly
    with_queue do |queue, consumer_n, _conn|
      cons = Pgque::Consumer.new(dsn, queue: queue, name: consumer_n,
                                 poll_interval: 10, logger: silent_logger)
      t = Thread.new { cons.start }
      sleep 0.5
      cons.stop
      finished = t.join(15)
      refute_nil finished, "consumer did not stop after stop()"
    end
  end

  def test_consumer_stop_returns_within_2s_while_waiting
    with_queue do |queue, consumer_n, _conn|
      cons = Pgque::Consumer.new(dsn, queue: queue, name: consumer_n,
                                 poll_interval: 30, logger: silent_logger)
      t = Thread.new { cons.start }
      sleep 1.0

      t0 = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      cons.stop
      t.join(5.0)
      elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t0

      refute t.alive?, "consumer thread did not stop"
      assert_operator elapsed, :<, 2.0,
                      "stop() took #{elapsed.round(2)}s; expected <2s"
    end
  end

  def test_consumer_wakes_on_pg_notify_before_poll_interval
    with_queue do |queue, consumer_n, conn|
      received = []
      received_evt = Mutex.new
      received_cv = ConditionVariable.new

      cons = Pgque::Consumer.new(dsn, queue: queue, name: consumer_n,
                                 poll_interval: 30, logger: silent_logger)
      cons.on("evt.wake") do |m|
        received_evt.synchronize do
          received << m
          received_cv.signal
        end
      end

      t = Thread.new { cons.start }
      sleep 1.5

      t_send = Process.clock_gettime(Process::CLOCK_MONOTONIC)
      producer = PG.connect(dsn)
      begin
        client = Pgque::Client.new(producer)
        client.send(queue, { "i" => 1 }, type: "evt.wake")
        producer.exec_params("select pgque.force_next_tick($1)", [queue])
        producer.exec_params("select pgque.ticker($1)", [queue])
        producer.exec_params("notify pgque_#{queue}, 'go'")
      ensure
        producer.close
      end

      woke = false
      received_evt.synchronize do
        deadline = Time.now + 5
        until received.any? || Time.now >= deadline
          received_cv.wait(received_evt, deadline - Time.now)
        end
        woke = received.any?
      end
      elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC) - t_send

      cons.stop
      t.join(5.0)

      assert woke, "consumer did not wake on pg_notify within 5s"
      assert_equal 1, received.size
      assert_operator elapsed, :<, 5.0,
                      "consumer woke too slowly (#{elapsed.round(2)}s)"
    end
  end

  def test_consumer_partial_batch_acks_good_messages_only
    with_queue do |queue, consumer_n, conn|
      client = Pgque::Client.new(conn)
      ok1 = client.send(queue, { "i" => 1 }, type: "ok")
      boom = client.send(queue, { "i" => 2 }, type: "boom")
      ok2 = client.send(queue, { "i" => 3 }, type: "ok")
      force_tick(conn, queue)

      seen_ok = []
      cons = Pgque::Consumer.new(dsn, queue: queue, name: consumer_n,
                                 poll_interval: 1, retry_after: 3600,
                                 logger: silent_logger)
      cons.on("ok") { |m| seen_ok << m.msg_id }
      cons.on("boom") { |_| raise "handler boom" }

      run_consumer_for(cons, 3.0).join(5.0)

      assert_includes seen_ok, ok1
      assert_includes seen_ok, ok2

      assert_operator retry_count_for_msg(conn, queue, boom), :>=, 1
      assert_equal 0, retry_count_for_msg(conn, queue, ok1)
      assert_equal 0, retry_count_for_msg(conn, queue, ok2)

      force_tick(conn, queue)
      follow = client.receive(queue, consumer_n, 10)
      ids = follow.map(&:msg_id)
      refute_includes ids, ok1
      refute_includes ids, ok2
      client.ack(follow[0].batch_id) if follow.any?
    end
  end
end
